src/Operators/diffusion.jl

```diff
@@ -0,0 +1,26 @@
+####
+#### Diffusive fluxes
+####
+
+@inline diffusive_flux_x(i, j, k, grid, c, κ_fcc) = κ_fcc * AxF(i, j, k, grid) * δx_faa(i, j, k, grid, c)
+@inline diffusive_flux_y(i, j, k, grid, c, κ_cfc) = κ_cfc * AyF(i, j, k, grid) * δy_afa(i, j, k, grid, c)
+@inline diffusive_flux_z(i, j, k, grid, c, κ_ccf) = κ_ccf *  Az(i, j, k, grid) * δz_aaf(i, j, k, grid, c)
+
+####
+#### Laplacian diffusion operator
+####
+
+"""
+    ∇κ∇c(i, j, k, grid, c, κ)
+
+Calculates diffusion for a tracer c via
+
+    1/V * [δx_caa(κ * Ax * δx_faa(c)) + δy_aca(κ * Ay * δy_afa(c)) + δz_aac(κ * Az * δz_aaf(c))]
+
+which will end up at the location `ccc`.
+"""
+@inline function ∇κ∇c(i, j, k, grid, c, κ)
+    1/VC(i, j, k, grid) * (δx_caa(i, j, k, grid, diffusive_flux_x, c, κ) +
```

Comment:
Should `VC` be `V_caa`? There are different volumes for each location?

Reply:
Hmmm, if the grid is only stretched in the vertical then `V_ccc = V_cfc = V_fcc = V_ffc`. I've used `VC` to mean a number of these, so I called it `V_aac`, although I don't think we use all four locations.

Maybe this is a bad idea? Certainly it will have to be changed if other dimensions are stretched as well.

Reply:
This function is agnostic to the type of grid. Thus to "future proof" the code for future non-uniform grids, we will need to use the correct volume here, despite the fact that a regular grid has only one volume, and a vertically stretched grid only has volumes that change along the vertical direction.

So, we need four volumes for tracer and the 3 velocity volumes.

But actually, the volume functions don't need to dispatch on the grid type; this can happen at a lower level correct? In other words we just need functions like

```julia
V_ccc(i, j, k, grid) = ΔxF(i, j, k, grid) * ΔyF(i, j, k, grid) * ΔzF(i, j, k, grid)
```

et cetera. Dispatching to the type of grid happens lower down.

Reply:
To clarify, I believe the four functions we need are

```julia
V_ccc(i, j, k, grid) = ΔxF(i, j, k, grid) * ΔyF(i, j, k, grid) * ΔzF(i, j, k, grid)
V_fcc(i, j, k, grid) = ΔxC(i-1, j, k, grid) * ΔyF(i, j, k, grid) * ΔzF(i, j, k, grid)
V_cfc(i, j, k, grid) = ΔxF(i, j, k, grid) * ΔyC(i, j-1, k, grid) * ΔzF(i, j, k, grid)
V_ccf(i, j, k, grid) = ΔxF(i, j, k, grid) * ΔyF(i, j, k, grid) * ΔzC(i, j, k-1, grid)
```

We can also define

```julia
const Vu = V_fcc
```

et cetera, as needed.

Reply:
So I've defined `Vᵃᵃᶜ` and `Vᵃᵃᶠ` as the two volumes we do use, but if you feel that we should be using three explicit superscripts then I'm happy to change it.

My argument for the any superscript is that I don't think we should define volume operators that aren't used with a vertically stretched grid because we have no way of testing them. E.g. `V_fcc` and `V_cfc` are the same on a vertically stretched grid.

Reply:
What do you think about my comment on dispatch at a lower-level?

If we could not dispatch at a low level then I feel the solution looks like

```julia
V_aac(i, j, k, grid::VerticallyStretchedGrid) = #
V_aaf(i, j, k, grid::VerticallyStretchedGrid) = #

V_ccc(i, j, k, grid::VerticallyStretchedGrid) = V_aac(i, j, k, grid)
V_fcc(i, j, k, grid::VerticallyStretchedGrid) = V_aac(i, j, k, grid)
V_cfc(i, j, k, grid::VerticallyStretchedGrid) = V_aac(i, j, k, grid)
V_ccf(i, j, k, grid::VerticallyStretchedGrid) = V_aaf(i, j, k, grid)
```

Then, flux derivatives can be written with grid-agnostic syntax.

Reply:
I'm not sure the comment about testing is an issue. We certainly can test that `V_ccc` is correct for a vertically stretched grid. We are not writing code that isn't used; we are creating a forward-looking framework. We *will* need this for more general non-uniform grids.